crucible-mir/src/Mir/Intrinsics.hs

```diff
@@ -801,6 +804,121 @@ readMirAggregateWithSymOffset bak iteFn off tpr ag@(MirAggregate totalSize m)
+    orPredBy f xs = do
+      case xs of
+        [] -> pure (falsePred sym)
+        (z : zs) -> do
+          zPred <- f z
+          zsPred <- mapM f zs
+          foldM (orPred sym) zPred zsPred
+
+    -- Whether `off` appears in the given `Run` of aggregate elements.
+    runPred :: Run -> IO (Pred sym)
+    runPred run = do
+      -- Note that we're using the unique stride as a proxy for element type
+      -- width, since we currently hardcode all aggregate entries as having
+      -- width 1 (TODO: hardcoded size=1). Once we move beyond that hardcoding,
+      -- we should use real entry widths here, rather than a proxy.
+      let tyWidth = rStride run
+
+      -- off >= rFrom
+      afterBeginning <- bvUge sym off =<< wordLit sym (rFrom run)
+
+      -- off < rTo (not `<=` because `rTo` is exclusive)
+      beforeEnd <- bvUlt sym off =<< wordLit sym (rTo run)
+
+      inBounds <- andPred sym afterBeginning beforeEnd
+
+      -- (off - rFrom) `mod` tyWidth == 0
+      relativeOff <- bvSub sym off =<< wordLit sym (rFrom run)
+      offModWidth <- bvUrem sym relativeOff =<< wordLit sym tyWidth
+      atTyBoundary <- bvEq sym offModWidth =<< wordLit sym 0
+
+      andPred sym inBounds atTyBoundary
+
+    -- Given a sorted list of element "spans", represented as @(fromOffset,
+    -- toOffset)@ pairs, find and return all contiguous sequences of two or more
+    -- elements of the same width, and the offsets all other elements.
+    foldRuns :: [(Word, Word)] -> ([Word], [Run])
+    foldRuns spans =
+      let (offsets, runs) = foldRuns' [] [] spans
+       in (reverse offsets, reverse runs)
+
+    foldRuns' :: [Word] -> [Run] -> [(Word, Word)] -> ([Word], [Run])
+    foldRuns' offsets runs spans =
+      case (runs, spans) of
+        -- Done
+        (_, []) ->
+          (offsets, runs)
+        -- Add to an existing run
+        (r : rs, s : ss)
+          | Just r' <- addToRun r s ->
+              foldRuns' offsets (r' : rs) ss
+        -- Add a new run
+        (_, s1 : s2 : ss)
+          | Just r <- mkRun s1 s2 ->
+              foldRuns' offsets (r : runs) ss
+        -- Add a new offset
+        (_, (sFrom, _) : ss) ->
+          foldRuns' (sFrom : offsets) runs ss
+
+    addToRun :: Run -> (Word, Word) -> Maybe Run
+    addToRun run (sFrom, sTo)
+      | rTo run == sFrom,
+        rStride run == sTo - sFrom =
+          Just (Run (rFrom run) sTo (rStride run))
+      | otherwise =
+          Nothing
+
+    mkRun :: (Word, Word) -> (Word, Word) -> Maybe Run
+    mkRun (s1From, s1To) s2 =
+      addToRun (Run s1From s1To (s1To - s1From)) s2
+
+-- | Represents a contiguous "run" of aggregate elements of the same width.
+-- Internal helper for `readMirAggregateWithSymOffset`.
+data Run = Run
+  { -- | Starting at (and inclusive of) this position
+    rFrom :: Word,
+    -- | Ending at (and exclusive of) this position
+    rTo :: Word,
+    -- | The spacing between the elements in this run
+    rStride :: Word
```

Comment:
Probably worth making these fields stricts to avoid unnecessary thunks:


```suggestion
  { -- | Starting at (and inclusive of) this position
    rFrom :: !Word,
    -- | Ending at (and exclusive of) this position
    rTo :: !Word,
    -- | The spacing between the elements in this run
    rStride :: !Word
```